src/app/icd/ICDManager.cpp

```diff
@@ -129,7 +126,15 @@ void ICDManager::UpdateICDMode()
             }
         }
     }
-    mICDMode = tempMode;
+
+    if (mICDMode != tempMode)
+    {
+        mICDMode = tempMode;
+        mStateObserverPool.ForEachActiveObject([](ObserverPointer * obs) {
```

Comment:
Instead of having loops everywhere when we want to call observer functions, what about a `PostEvent(enum icd_event_type event)` method that does the looping and calls the appropriate Observer methods depending on the event parameter?